BUG: fixing regex to avoid DoS vulnerabilities

nealchau opened the pull request:
Regexes which can match over long parts of strings but then
be rejected by nonmatching at the end can be used for
denial-of-service.  This addresses #17255 and #17254.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

Commit:
BUG: fixing regex to avoid DoS vulnerabilities

Regexes which can match over long strings but then
be rejected by nonmatching at the end can be used for
denial-of-service.  This addresses #17255 and #17254.

eric-wieser commented on numpy/distutils/conv_template.py:
```diff
@@ -152,7 +152,7 @@ def parse_values(astr):
 
 stripast = re.compile(r"\n\s*\*?")
-named_re = re.compile(r"#\s*(\w*)\s*=([^#]*)#")
+named_re = re.compile(r"#\s*([\S\w]*)\s*=([^#]*)#")
```
This changes the meaning of the regex to match all non-space characters, which isn't something we want to do. Can you explain why this change makes any difference?

yetingli replied:
i think you can replace `\s*(\w*)\s*`  with `\s*(\w+\s*)? `

The latter is safe

eric-wieser commented on numpy/distutils/from_template.py:
```diff
@@ -83,8 +83,8 @@ def parse_structure(astr):
-named_re = re.compile(r"<\s*(\w[\w\d]*)\s*=\s*(.*?)\s*>")
-list_re = re.compile(r"<\s*((.*?))\s*>")
+named_re = re.compile(r"<\s*(\w[\w\d]*)\s*=\s*(\S*?)\s*>")
+list_re = re.compile(r"<\s*((\S*?))\s*>")
```
These two changes look correct, but I don't understand why they makes any difference.

yetingli replied:
This sub-pattern `\s*(\S*?)\s*` will still produce catastrophic backtracking. 
i think you can replace `\s*(\S*?)\s*` with `\s*(\S+\s*)?`

nealchau replied:
Thanks @yetingli yes that makes sense, I just updated my PR and also added comments using re.VERBOSE.  The goal is to make the require less backtracking and include the termination within the regex as it moves forward instead of matching arbitrarily long strings and then finding a mismatch afterwards that requires backtracking.

Commit:
BUG: fix and describe regexes to avoid DoS

Here I updated the regexes per @yetingli's comments
and also added comments within using the re.VERBOSE feature.
The goal is to make the regex know when the expression
has matched or not without introducing unnecessary backtracking
which can take excessive time to evaluate per #17255 and #17254.

Commit:
BUG: fixing reges to avoid DoS vulnerabilities

removed an extra quote from the multiline quote

mattip commented:
It might help debug the CI failures if there were test added to `numpy/distutils/tests/test_from_template.py`

rgommers commented:
As @eric-wieser pointed out in https://github.com/numpy/numpy/issues/17254#issuecomment-687572810, this change is unnecessary because arbitrary code execution is already possible for the user accessing this piece of `numpy.distutils`.

So I propose to close this PR. Thanks for contributing though @nealchau, much appreciated.

seberg commented:
OK, I will close this, thank you for the PR.

I don't mind updating the regex, but right now its broken (probably requires non-capturing groups here).  So feel free to open a new PR if you want. Before fixing it this way, I would appreciate if you can check whether `\w+` instead of `\w*` is not the correct regex to begin with though.
